File: app/src/main/java/com/rucola/app/domain/UseCases.kt
package com.rucola.app.domain

import com.rucola.app.data.RucolaRepository

fun validateMessage(type: MessageType, body: String, mediaReference: String? = null) {
    when (type) {
        MessageType.TEXT -> require(body.isNotBlank()) { "Text messages need a message body" }
        MessageType.EMOJI -> require(body.isNotBlank()) { "Emoji messages need an emoji" }
        MessageType.PHOTO_VIDEO, MessageType.DRAWING -> require(body.isNotBlank() || !mediaReference.isNullOrBlank()) {
            "Media messages need text or a media reference"
        }
    }
}

class SaveSetup(private val repository: RucolaRepository) {
    suspend operator fun invoke(
        partnerNickname: String,
        ownName: String,
        partnerColor: String,
        togetherSince: Long?,
    ) = repository.saveSetup(partnerNickname, ownName, partnerColor, togetherSince)
}

class SendMessage(private val repository: RucolaRepository) {
    suspend operator fun invoke(type: MessageType, body: String, mediaReference: String? = null) {
        validateMessage(type, body, mediaReference)
        repository.sendMessage(type, body, mediaReference)
    }
}

File: src/domain/useCases.ts
import type { MessageType, Participant } from './models';
import type { RucolaRepository } from './repository';

export class SaveSetup {
  constructor(private readonly repository: RucolaRepository) {}

  execute(input: {
    partnerNickname: string;
    ownName: string;
    togetherSince: number | null;
    partnerColor?: string;
  }): Promise<void> {
    return this.repository.saveSetup(input);
  }
}

export class SendMessage {
  constructor(private readonly repository: RucolaRepository) {}

  execute(input: {
    type: MessageType;
    body: string;
    mediaReference?: string | null;
  }) {
    return this.repository.sendMessage(input);
  }
}

export class GetActiveMessage {
  constructor(private readonly repository: RucolaRepository) {}

  execute(participant: Participant) {
    return this.repository.getActiveMessage(participant);
  }
}

export class GetHistory {
  constructor(private readonly repository: RucolaRepository) {}

  async execute() {
    const messages = await this.repository.getMessages();
    return messages.filter((message) => !message.isActive);
  }
}
